File: Maximum_Product_Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        N = len(nums)
        maxi = nums[0]
        inter_max = nums[0]  # keeping track of the minimum product
        inter_min = nums[0]  # keeping track of the maximum product
        for i in range(1, N):
            inter_max, inter_min = max(nums[i], nums[i]*inter_max, nums[i]*inter_min), min(nums[i], nums[i]*inter_min, nums[i]*inter_max)
            maxi = max(inter_max, maxi)


        return maxi

File: Maximum_Product_Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {


        const int& N = nums.size();
        int prod = nums[0];
        int min_prod = nums[0];
        int max_prod = nums[0];
        int maxi = nums[0];
        for (int i = 1; i < N; i++) {

            int temp = max_prod;
            max_prod = max(nums[i], max(nums[i] * min_prod, nums[i] * temp));
            min_prod = min(nums[i], min(nums[i] * min_prod, nums[i] * temp));
            



            maxi = max(maxi, max_prod);




        }
        return maxi;

    }
};
